### ISSUE TITLE
Phone numbers field bug: leading zeros removed

### ISSUE THREAD

**Issue by Lyquix** [FLEXIcontent/flexicontent-cck#430]
I received a report of the phone numbers field removing leading zeros from the phone number. For example, the phone number 215-900-0099 would be rendered as 215-900-99.

The cause of this bug is found in the validation done before save, here: 
https://github.com/FLEXIcontent/flexicontent-cck/blob/master/plugins/flexicontent_fields/phonenumbers/phonenumbers.php#L450

```
$newpost[$new]['phone1'] = !strlen($v['phone1']) ? '' : flexicontent_html::dataFilter(@$v['phone1'], $phone_maxlength, 'INT', 0);
$newpost[$new]['phone2'] = !strlen($v['phone2']) ? '' : flexicontent_html::dataFilter(@$v['phone2'], $phone_maxlength, 'INT', 0);
$newpost[$new]['phone3'] = !strlen($v['phone3']) ? '' : flexicontent_html::dataFilter(@$v['phone3'], $phone_maxlength, 'INT', 0);
```

Changing the validation to STRING works but may be too permissive, so I think it would be best to create some custom filtering that allows only numbers (0-9) and uppercase letters (A-Z), for example:

```
preg_replace("/[^0-9A-Z]/", '', $string);
```

We can then pass the result through `flexicontent_html::dataFilter` as STRING, but I am not sure if that is necessary (except for enforcing max length, but we are current not imposing any length limit).

Please let me know if you agree and I can make a commit that I have ready.

**Comment by Lyquix** [comment]
This is what I had in mind:

```
// Remove any characters that aren't a digit or uppercase letter A-Z

$v['phone1'] = preg_replace('/[^0-9A-Z]/', '', strtoupper($v['phone1']));
$v['phone2'] = preg_replace('/[^0-9A-Z]/', '', strtoupper($v['phone2']));
$v['phone3'] = preg_replace('/[^0-9A-Z]/', '', strtoupper($v['phone3']));

// Validate phone number, skipping phone number that are empty after validation

$newpost[$new]['phone1'] = !strlen($v['phone1']) ? '' : flexicontent_html::dataFilter(@$v['phone1'], $phone_maxlength, 'STRING', 0);
$newpost[$new]['phone2'] = !strlen($v['phone2']) ? '' : flexicontent_html::dataFilter(@$v['phone2'], $phone_maxlength, 'STRING', 0);
$newpost[$new]['phone3'] = !strlen($v['phone3']) ? '' : flexicontent_html::dataFilter(@$v['phone3'], $phone_maxlength, 'STRING', 0);
```

And remove the class `validate-numeric` from the input fields so that non-numeric characters are not blocked in the item form.

**Comment by ggppdk** [comment]
We must have validation. But Validation can be custom of course. Anything we need as long we do not allow anything dangerous eg JS script code. So the code  is 100% acceptable. But it is best to have a parameter allow only number or number and Latin letter too

**Comment by ggppdk** [comment]
And we do not need to call the datafilter function if we use custom validation and if we also limit the character length. The datafilter function is just reusable code

**Comment by Lyquix** [comment]
Ok, great idea, I will make a commit that includes:

* New parameter to control whether letters are allowed or not
* Code to add/remove CSS class validate-numeric for Javascript validation
* Regex for validation changes based on parameter
* Enforce max length on each phoneX component

**Comment by ggppdk** [comment]
Despite of using any validation we need, if some case matches any of those supported by data Filter
- then you can use it
e.g **'ALNUM'** is
 A-Z (cas-insensitive) and numbers 0-1

so you can use 'ALNUM' and then  **strtoupper($v['phone1'])**  too

dataFilter you can be passed:
**1**:  SAFEHTML
**2**: Joomla TEXT Filtering for current user
**'EMAIL'**:  ... supporting punnycode
**'URL'**: ... internal AND **EXTERNAL** URLs (supporting punnycode)

and then all that are supported by JFilterInput::clean()

INT:       An integer,
UINT:      An unsigned integer,
FLOAT:     A floating point number,
BOOLEAN:   A boolean value,
WORD:      A string containing A-Z or underscores only (not case sensitive),
ALNUM:     A string containing A-Z or 0-9 only (not case sensitive),
CMD:       A string containing A-Z, 0-9, underscores, periods or hyphens (not case sensitive),
BASE64:    A string containing A-Z, 0-9, forward slashes, plus or equals (not case sensitive),
STRING:    A fully decoded and sanitised string (default),
HTML:      A sanitised string,
ARRAY:     An array,
PATH:      A sanitised file path,
TRIM:      A string trimmed from normal, non-breaking and multibyte spaces
USERNAME:  Do not use (use an application specific filter),
RAW:       The raw string is returned with no filtering,
unknown:   An unknown filter will act like STRING. If the input is an array it will return an
           array of fully decoded and sanitised strings.

**Comment by Lyquix** [comment]
Ok I will use the ALNUM filter when we allow letters. I will make a commit today.

**Comment by ggppdk** [comment]
There is no need since current code is good
Fixed with 36ae6a5
